src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs

```diff
@@ -130,24 +177,43 @@ public virtual bool VerifyData(Stream data, byte[] signature, HashAlgorithmName
             if (signature == null) { throw new ArgumentNullException(nameof(signature)); }
             if (string.IsNullOrEmpty(hashAlgorithm.Name)) { throw HashAlgorithmNameNullOrEmpty(); }
 
-            byte[] hash = HashData(data, hashAlgorithm);
-            return VerifySignature(hash, signature);
+            return VerifyDataCore(data, signature, hashAlgorithm, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);
         }
 
-        public virtual bool TryCreateSignature(ReadOnlySpan<byte> hash, Span<byte> destination, out int bytesWritten)
+        public byte[] CreateSignature(byte[] rgbHash, DSASignatureFormat signatureFormat)
         {
-            byte[] sig = CreateSignature(hash.ToArray());
-            if (sig.Length <= destination.Length)
+            if (rgbHash == null)
             {
-                new ReadOnlySpan<byte>(sig).CopyTo(destination);
-                bytesWritten = sig.Length;
-                return true;
-            }
-            else
-            {
-                bytesWritten = 0;
-                return false;
+                throw new ArgumentNullException(nameof(rgbHash));
             }
+
+            return CreateSignatureCore(rgbHash, signatureFormat);
+        }
+
+        protected virtual byte[] CreateSignatureCore(ReadOnlySpan<byte> hash, DSASignatureFormat signatureFormat)
+        {
+            int size = GetMaxSignatureSize(signatureFormat);
+            Debug.Assert(size <= 256, "GetMaxSignatureSize returned more than expected");
+            Span<byte> signature = stackalloc byte[size];
+            bool result = TryCreateSignatureCore(hash, signature, signatureFormat, out int bytesWritten);
+            Debug.Assert(result, "TryCreateSignatureCore unexpectedly returned false");
+            return signature.Slice(0, bytesWritten).ToArray();
+        }
+
+        public virtual bool TryCreateSignature(ReadOnlySpan<byte> hash, Span<byte> destination, out int bytesWritten)
+            => TryCreateSignatureCore(hash, destination, DSASignatureFormat.IeeeP1363FixedFieldConcatenation, out bytesWritten);
+
+        public bool TryCreateSignature(ReadOnlySpan<byte> hash, Span<byte> destination, DSASignatureFormat signatureFormat, out int bytesWritten)
+            => TryCreateSignatureCore(hash, destination, signatureFormat, out bytesWritten);
+
+        protected virtual bool TryCreateSignatureCore(ReadOnlySpan<byte> hash, Span<byte> destination, DSASignatureFormat signatureFormat, out int bytesWritten)
+        {
+            // This method is expected to be overriden with better implementation
```

Comment:
When we update the docs for this class will we recommend falling back to this implementation for an unknown signature format, since we'd theoretically handle it in ConvertIeeeP1363Signature for them?

Alternatively, do we want to add something so a derived type can report whether or not it directly handles a format, so that we understand when to apply an 1363-to-destination-scheme strategy?

Reply:
(Maybe it's moot since there're only the two, but I like to think ahead)